apps/service_providers/migrations/0019_auto_20241029_0901.py

```diff
@@ -0,0 +1,222 @@
+            team__isnull=True,
+            type=assistant.llm_provider.type,
+            name=assistant.llm_model,
+        )
+        assistant.llm_provider_model = global_provider_model
+        assistant.save()
+        return
+    except LlmProviderModel.DoesNotExist:
+        pass
+
+    try:
+        custom_provider_model = LlmProviderModel.objects.get(
+            team=assistant.team,
+            type=assistant.llm_provider.type,
+            name=assistant.llm_model,
+        )
+        assistant.llm_provider_model = custom_provider_model
+        assistant.save()
+        return
+    except LlmProviderModel.DoesNotExist:
+        pass
+
+    new_custom_provider_model = LlmProviderModel.objects.create(
+        team=assistant.team,
+        type=assistant.llm_provider.type,
+        name=assistant.llm_model,
+    )
+    assistant.llm_provider_model = new_custom_provider_model
+    assistant.save()
+
+def _handle_llm_experiment(LlmProviderModel, experiment):
+    try:
+        global_provider_model = LlmProviderModel.objects.get(
+            team__isnull=True,
+            type=experiment.llm_provider.type,
+            name=experiment.llm,
+            max_token_limit=experiment.max_token_limit
+        )
+        experiment.llm_provider_model = global_provider_model
+        experiment.save()
+        return
+    except LlmProviderModel.DoesNotExist:
+        pass
+
+    try:
+        custom_provider_model = LlmProviderModel.objects.get(
+            team=experiment.team,
+            type=experiment.llm_provider.type,
+            name=experiment.llm,
+            max_token_limit=experiment.max_token_limit
+        )
+        experiment.llm_provider_model = custom_provider_model
+        experiment.save()
+        return
+    except LlmProviderModel.DoesNotExist:
+        pass
+
+    new_custom_provider_model = LlmProviderModel.objects.create(
+        team=experiment.team,
+        type=experiment.llm_provider.type,
+        name=experiment.llm,
+        max_token_limit=experiment.max_token_limit,
+    )
+    experiment.llm_provider_model = new_custom_provider_model
+    experiment.save()
+
+def _create_custom_llm_provider_models(apps, schema_editor):
+    LlmProvider = apps.get_model("service_providers", "LlmProvider")
+    LlmProviderModel = apps.get_model("service_providers", "LlmProviderModel")
+    Experiment = apps.get_model("experiments", "Experiment")
+    Analysis = apps.get_model("analysis", "Analysis")
+    Node = apps.get_model("pipelines", "Node")
+
+    for analysis in Analysis.objects.select_related("llm_provider").all():
+        _handle_analysis(LlmProviderModel, analysis)
+
+    for node in Node.objects.filter(Q(type="LLMResponseWithPrompt") | Q(type="LLMResponse") | Q(type="RouterNode")).all():
+        _handle_pipeline_node(LlmProvider, LlmProviderModel, node)
+
+    for experiment in Experiment.objects.select_related("llm_provider").all():
+        if experiment.llm_provider is None and experiment.assistant:
```

Comment:
I think the `experiment.llm_provider is None` check should be dropped. Reason being that we don't remove the llm provider from the experiment instance when we link an assistant, so this check will essentially exclude those assistants (there are 16 cases like this on prod). 0 cases for pipelines.

Reply:
https://github.com/dimagi/open-chat-studio/pull/810/commits/b82b3ecd55030ce39fe590b7b8f6c4f2e79faa6d